File: scripts/lib/technology-image-backup.mjs
/**
 * Backup assets/images/technology files before delete or overwrite.
 * Dest: assets/images/technology/backup/{relative-dir}/{name}.{timestamp}{ext}
 */
import { copyFileSync, existsSync, mkdirSync, unlinkSync } from 'fs';
import { dirname, join, relative, resolve, isAbsolute } from 'path';
import { fileURLToPath } from 'url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..', '..');
export const TECH_IMG_DIR = join(ROOT, 'assets', 'images', 'technology');

function timestampLocal() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, '0');
  return `${d.getFullYear()}${p(d.getMonth() + 1)}${p(d.getDate())}-${p(d.getHours())}${p(d.getMinutes())}${p(d.getSeconds())}`;
}

/** @returns {boolean} */
export function isUnderTechnologyImages(absPath) {
  const abs = resolve(absPath);
  const techRoot = resolve(TECH_IMG_DIR);
  const rel = relative(techRoot, abs);
  return rel === '' || (!rel.startsWith('..') && !isAbsolute(rel));
}

/**
 * Copy file into technology/backup/ preserving subpath (source/, reviewed/, etc.).
 * @returns {string | null} backup absolute path
 */
export function backupTechnologyImage(absPath, { reason = '', dryRun = false } = {}) {
  if (!existsSync(absPath)) return null;

  const abs = resolve(absPath);
  const techRoot = resolve(TECH_IMG_DIR);
  if (!isUnderTechnologyImages(abs)) {
    throw new Error(`Not under assets/images/technology: ${absPath}`);
  }

  const rel = relative(techRoot, abs).replace(/\\/g, '/');
  if (rel.startsWith('backup/')) return null;

  const slash = rel.lastIndexOf('/');
  const dirPart = slash >= 0 ? rel.slice(0, slash) : '';
  const file = slash >= 0 ? rel.slice(slash + 1) : rel;
  const dot = file.lastIndexOf('.');
  const stem = dot >= 0 ? file.slice(0, dot) : file;
  const ext = dot >= 0 ? file.slice(dot) : '';
  const backedName = `${stem}.${timestampLocal()}${ext}`;
  const dest = dirPart
    ? join(techRoot, 'backup', dirPart, backedName)
    : join(techRoot, 'backup', backedName);

  if (dryRun) {
    console.log(
      `[dry-run] backup ${rel} → backup/${dirPart ? dirPart + '/' : ''}${backedName}${reason ? ` (${reason})` : ''}`,
    );
    return dest;
  }

  mkdirSync(dirname(dest), { recursive: true });
  copyFileSync(abs, dest);
  const relDest = relative(techRoot, dest).replace(/\\/g, '/');
  console.log(`Backed up: ${rel} → ${relDest}${reason ? ` (${reason})` : ''}`);
  return dest;
}

/** Backup then delete. @returns {string | null} backup path */
export function backupAndDeleteTechnologyImage(absPath, opts = {}) {
  const dest = backupTechnologyImage(absPath, opts);
  if (opts.dryRun) {
    if (existsSync(absPath)) {
      const rel = relative(TECH_IMG_DIR, resolve(absPath)).replace(/\\/g, '/');
      console.log(`[dry-run] delete ${rel}`);
    }
    return dest;
  }
  if (existsSync(absPath)) unlinkSync(absPath);
  return dest;
}

File: scripts/lib/technology_image_backup.py
"""Backup assets/images/technology files before delete (mirrors technology-image-backup.mjs)."""
from __future__ import annotations

import shutil
from datetime import datetime
from pathlib import Path


def _timestamp_local() -> str:
    return datetime.now().strftime("%Y%m%d-%H%M%S")


def backup_technology_image(
    file_path: Path,
    img_dir: Path,
    *,
    reason: str = "",
    dry_run: bool = False,
) -> Path | None:
    file_path = file_path.resolve()
    img_dir = img_dir.resolve()
    if not file_path.is_file():
        return None
    try:
        rel = file_path.relative_to(img_dir)
    except ValueError:
        return None
    rel_posix = rel.as_posix()
    if rel_posix.startswith("backup/"):
        return None

    stamp = _timestamp_local()
    dest = img_dir / "backup" / rel.parent / f"{file_path.stem}.{stamp}{file_path.suffix}"
    suffix = f" ({reason})" if reason else ""
    if dry_run:
        print(f"[dry-run] backup {rel_posix} → {dest.relative_to(img_dir).as_posix()}{suffix}")
        return dest

    dest.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(file_path, dest)
    print(f"Backed up: {rel_posix} → {dest.relative_to(img_dir).as_posix()}{suffix}")
    return dest


def backup_and_unlink(
    file_path: Path,
    img_dir: Path,
    *,
    reason: str = "",
    dry_run: bool = False,
) -> Path | None:
    dest = backup_technology_image(file_path, img_dir, reason=reason, dry_run=dry_run)
    if dry_run:
        if file_path.is_file():
            try:
                rel = file_path.relative_to(img_dir)
                print(f"[dry-run] delete {rel.as_posix()}")
            except ValueError:
                pass
        return dest
    if file_path.is_file():
        file_path.unlink(missing_ok=True)
    return dest
